scipy/optimize/_equality_constrained_sqp/equality_constrained_sqp.py

```diff
@@ -0,0 +1,256 @@
+"""Byrd-Omojokun Trust-Region SQP method."""
+
+from __future__ import division, print_function, absolute_import
+import scipy.sparse as spc
+from .projections import projections
+from .qp_subproblem import modified_dogleg, projected_cg, box_intersections
+import numpy as np
+from numpy.linalg import norm
+
+__all__ = ['equality_constrained_sqp']
+
+
+def default_scaling(x):
+    n, = np.shape(x)
+    return spc.eye(n)
+
+
+def equality_constrained_sqp(fun, grad, hess, constr, jac,
+                             x0, fun0, grad0, constr0,
+                             jac0, stop_criteria, state,
+                             trust_lb=None,
+                             trust_ub=None,
+                             initial_penalty=1.0,
+                             initial_trust_radius=1.0,
+                             scaling=default_scaling,
+                             return_all=False,
+                             factorization_method=None):
+    """Solve nonlinear equality-constrained problem using trust-region SQP.
+
+    Solve optimization problem:
+
+        minimize fun(x)
+        subject to: constr(x) = 0
+
+    using Byrd-Omojokun Trust-Region SQP method described in [1]_. Several
+    implementation details are based on [2]_ and [3]_, p. 549.
+
+    References
+    ----------
+    .. [1] Lalee, Marucha, Jorge Nocedal, and Todd Plantenga. "On the
+           implementation of an algorithm for large-scale equality
+           constrained optimization." SIAM Journal on
+           Optimization 8.3 (1998): 682-706.
+    .. [2] Byrd, Richard H., Mary E. Hribar, and Jorge Nocedal.
+           "An interior point algorithm for large-scale nonlinear
+           programming." SIAM Journal on Optimization 9.4 (1999): 877-900.
+    .. [3] Nocedal, Jorge, and Stephen J. Wright. "Numerical optimization"
+           Second Edition (2006).
+    """
+    PENALTY_FACTOR = 0.3  # Rho from formula (3.51), reference [2]_, p.891.
```

Comment:
Do you think I should allow the user to set these parameters, and include some documentation about them in the ``minimize_constrained`` interface?